x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/profile/TransportProfileHasPrivilegesAction.java

```diff
@@ -0,0 +1,125 @@
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.SecurityContext;
+import org.elasticsearch.xpack.core.security.action.user.ProfileHasPrivilegesAction;
+import org.elasticsearch.xpack.core.security.action.user.ProfileHasPrivilegesRequest;
+import org.elasticsearch.xpack.core.security.action.user.ProfileHasPrivilegesResponse;
+import org.elasticsearch.xpack.core.security.authc.Subject;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.privilege.ApplicationPrivilegeDescriptor;
+import org.elasticsearch.xpack.security.authz.AuthorizationService;
+import org.elasticsearch.xpack.security.authz.store.NativePrivilegeStore;
+import org.elasticsearch.xpack.security.profile.ProfileService;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+public class TransportProfileHasPrivilegesAction extends HandledTransportAction<ProfileHasPrivilegesRequest, ProfileHasPrivilegesResponse> {
+
+    private static final Logger logger = LogManager.getLogger(TransportProfileHasPrivilegesAction.class);
+
+    private final AuthorizationService authorizationService;
+    private final NativePrivilegeStore privilegeStore;
+    private final ProfileService profileService;
+    private final SecurityContext securityContext;
+    private final ThreadPool threadPool;
+
+    @Inject
+    public TransportProfileHasPrivilegesAction(
+        TransportService transportService,
+        ActionFilters actionFilters,
+        AuthorizationService authorizationService,
+        NativePrivilegeStore privilegeStore,
+        ProfileService profileService,
+        SecurityContext securityContext,
+        ThreadPool threadPool
+    ) {
+        super(ProfileHasPrivilegesAction.NAME, transportService, actionFilters, ProfileHasPrivilegesRequest::new);
+        this.authorizationService = authorizationService;
+        this.privilegeStore = privilegeStore;
+        this.profileService = profileService;
+        this.securityContext = securityContext;
+        this.threadPool = threadPool;
+    }
+
+    @Override
+    protected void doExecute(Task task, ProfileHasPrivilegesRequest request, ActionListener<ProfileHasPrivilegesResponse> listener) {
+        profileService.getProfileSubjects(request.profileUids(), ActionListener.wrap(profileSubjectsAndFailures -> {
+            if (profileSubjectsAndFailures.profileUidToSubject().isEmpty()) {
+                listener.onResponse(new ProfileHasPrivilegesResponse(Set.of(), Set.of()));
```

Comment:
```suggestion
                listener.onResponse(new ProfileHasPrivilegesResponse(Set.of(), Set.copyOf(profileUidToSubject.failureProfileUids)));
```
?

Reply:
This is a nice catch, pushed [62cfa87](https://github.com/elastic/elasticsearch/pull/85898/commits/62cfa8712f618e7bde7b3f4b5683d7c6b5f974ba), which includes a test.